### PR TITLE
Add some additional checks to the acceptance mempool test

### LINKED ISSUE #2691: Add `zebrad` acceptance tests for the mempool
## Motivation

The mempool unit tests cover a lot of mempool functionality, using `MockService`s where needed.

But there are some things we can't test unless we actually run `zebrad`:
- [x] mempool startup from `zebrad`'s `start` command
    - the unit tests already check that the mempool starts disabled
- [x] mempool startup while the rest of `zebrad` is running (covered by PR #2862)
- [ ] mempool running while the rest of `zebrad` is also running
- [ ] The mempool doesn't activate really early (while we're syncing the first few checkpoints)

We've had panics and deadlocks in similar situations before. (Particularly the start command & running everything at the same time.)

So it's important that we run some tests like this.

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We want to check some additional stuff to the mempool acceptance test we added at https://github.com/ZcashFoundation/zebra/pull/2862

Close https://github.com/ZcashFoundation/zebra/issues/2691

## Solution

Add missing checks.

## Review

The code added here is short, so anyone can review.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebrad/tests/acceptance.rs [hunk 1]
```diff
@@ -781,7 +781,7 @@ fn activate_mempool_mainnet() -> Result<()> {
         Height(1),
         Mainnet,
         STOP_AT_HEIGHT_REGEX,
-        SMALL_CHECKPOINT_TIMEOUT,
+        LARGE_CHECKPOINT_TIMEOUT,
```
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs]: I added more time here as i want to make sure the we can verify up to the block 400 when the mempool is active.

File: zebrad/tests/acceptance.rs [hunk 2]
```diff
@@ -887,10 +887,22 @@ fn sync_until(
     if enable_mempool_at_height.is_some() {
         child.expect_stdout_line_matches("enabling mempool for debugging")?;
         child.expect_stdout_line_matches("activating mempool")?;
+
+        // make sure zebra is running with the mempool
+        child.expect_stdout_line_matches("verified checkpoint range block_count=400")?;
```
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs]: This is testing that the zebrad is running with the mempool at the same time.

File: zebrad/tests/acceptance.rs [hunk 3]
```diff
@@ -887,10 +887,22 @@ fn sync_until(
...
 
     child.expect_stdout_line_matches(stop_regex)?;
-    child.kill()?;
+
+    // make sure there is never a mempool if we don't explicity enable it
+    if enable_mempool_at_height.is_none() {
+        // this call failure will kill the process
+        assert!(child
+            .expect_stdout_line_matches("activating mempool")
+            .is_err());
```
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs]: This is making sure that no mempool will be running early in any of our current tests that use `sync_until` unless is specified.

File: zebrad/tests/acceptance.rs (hunk 1 above)
Comment (teor2345) [reply zebrad/tests/acceptance.rs]: Test Purpose: `activate_mempool_mainnet` activates the mempool, then exits.

In this test, the state makes Zebra exit as soon as it commits block `Height(1)`.
So the mempool won't run for very long.

Can we make a separate "running mempool" test that syncs up to `LARGE_CHECKPOINT_TEST_HEIGHT`?

File: zebrad/tests/acceptance.rs (hunk 2 above)
Comment (teor2345) [reply zebrad/tests/acceptance.rs]: Test Reliability: Is the exact block count important to this test?

Sometimes Zebra will commit two checkpoint ranges at the same time, so the block count would be 800.

```suggestion
        child.expect_stdout_line_matches("verified checkpoint range")?;
```

File: zebrad/tests/acceptance.rs [hunk 4]
```diff
@@ -887,10 +887,22 @@ fn sync_until(
...
+    } else {
+        child.kill()?;
+    }
```
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Future bug risk: What if we add more code here, and forget to kill the child process?

It's ok to try to kill the child process multiple times, or if it has already exited.

```suggestion
    } 
    
    child.kill()?;
```

File: zebrad/tests/acceptance.rs (hunk 3 above)
Comment (teor2345) [reply zebrad/tests/acceptance.rs]: Likely Test Bug: The child process isn't killed if the test fails.

If the mempool is activated, the assertion fails, which kills the test process.
But I don't see anything that kills the `zebrad` child process.
So does it just keep running?

```suggestion
    if enable_mempool_at_height.is_none() {
        // if there is no matching line, the `expect_stdout_line_matches` error kills the `zebrad` child.
        // the error is delayed until the test timeout, or until the child reaches the stop height and exits.
        let mempool_is_activated = child
            .expect_stdout_line_matches("activating mempool")
            .is_ok();
            
        // if there is a matching line, we panic and kill the test process.
        // but we also need to kill the `zebrad` child before the test panics.
        if (mempool_is_activated) {
            child.kill()?;
            panic!("unexpected mempool activation: mempool should not activate while syncing lots of blocks")
        }
```

**Review by teor2345 — COMMENTED** [review]
This is a good idea for a test.

But I'd like to create a new test, rather than re-using the existing one.
A new test could sync more blocks, and get better test coverage.

I also made some suggestions for making the test more reliable.

File: zebrad/tests/acceptance.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: I didnt realized about this but you are right. Thanks. Added an additional test at https://github.com/ZcashFoundation/zebra/pull/2880/commits/ca88682b0e7e18092086cbfd39689bffb83c4a7a

File: zebrad/tests/acceptance.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Yea, that should be good enough. Done as part of https://github.com/ZcashFoundation/zebra/pull/2880/commits/ca88682b0e7e18092086cbfd39689bffb83c4a7a

File: zebrad/tests/acceptance.rs (hunk 4 above)
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Done this as part of https://github.com/ZcashFoundation/zebra/pull/2880/commits/31f736366ec2dd36c27cea13d35af875ca5cf02c

File: zebrad/tests/acceptance.rs (hunk 3 above)
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: This was done at https://github.com/ZcashFoundation/zebra/pull/2880/commits/31f736366ec2dd36c27cea13d35af875ca5cf02c

File: zebrad/tests/acceptance.rs [hunk 5]
```diff
@@ -887,9 +903,28 @@ fn sync_until(
...
+            .is_ok();
+
+        // if there is a matching line, we panic and kill the test process.
+        // but we also need to kill the `zebrad` child before the test panics.
+        if mempool_is_activated {
+            child.kill()?;
+            panic!("unexpected mempool activation: mempool should not activate while syncing lots of blocks")
+        }
+    }
     child.kill()?;
```
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: I was wrong about calling `kill` multiple times.

Here's the fix to ignore the "can't kill an exited process" error.

```suggestion

    // make sure the child process is dead
    // if it has already exited, ignore that error
    let _ = child.kill();
```

**Review by teor2345 — APPROVED** [review]
I just committed a minor fix, we should be good to merge now

**Comment by teor2345** [comment]
I'm auto-merging this PR, so upcoming PRs are better tested.

Thanks!

**Pull request merged**